node/core/runtime-api/src/lib.rs

```diff
@@ -335,12 +335,52 @@ where
 {
 	let _timer = metrics.time_make_runtime_api_request();
 
+	// The version of the `ParachainHost` runtime API that we are absolutely sure is deployed widely
+	// on all supported networks: Rococo, Kusama, Polkadot and perhaps others like Versi.
+	//
+	// This version is used for eliding unnecessary runtime API calls. It is safer to keep the lower
+	// version.
+	const WIDELY_DEPLOYED_API_VERSION: u32 = 1;
+
 	macro_rules! query {
-		($req_variant:ident, $api_name:ident ($($param:expr),*), $sender:expr) => {{
+		($req_variant:ident, $api_name:ident ($($param:expr),*), ver = $version:literal, $sender:expr) => {{
 			let sender = $sender;
 			let api = client.runtime_api();
-			let res = api.$api_name(&BlockId::Hash(relay_parent) $(, $param.clone() )*)
-				.map_err(|e| RuntimeApiError::from(format!("{:?}", e)));
+
+			let runtime_version = if $version > WIDELY_DEPLOYED_API_VERSION {
+				use sp_api::ApiExt;
+				api.api_version::<dyn ParachainHost<Block>>(&BlockId::Hash(relay_parent))
+					.unwrap_or_else(|e| {
+						tracing::warn!(
+							target: LOG_TARGET,
+							"cannot query the runtime API version: {}",
+							e,
+						);
+						Some(WIDELY_DEPLOYED_API_VERSION)
+					})
+					.unwrap_or_else(|| {
+						tracing::warn!(
+							target: LOG_TARGET,
+							"no runtime version is reported"
+						);
+						WIDELY_DEPLOYED_API_VERSION
+					})
+			} else {
+				// The required version is less or equal to the widely deployed runtime API version.
+				WIDELY_DEPLOYED_API_VERSION
+			};
```

Comment:
FYI, we are caching the runtime version. So, getting the runtime version does not involve calling into the runtime

Reply:
I actually considered this. My reasoning was that runtime APIs can be called on different blocks. But writing this I realize that this may be not a good reason. 

- We should not worry about the runtime cache evictions that much: in any case there won't be more than two active runtimes, no matter how many heads there are.
- There is no reason for trying to minimize cases where we query runtime version, because it will still be called every block by some subsystem.

Well, I guess we can simplify that!

Reply:
There is also tons of other stuff that will call into the runtime on every block. So, not that much to worry about :P